### PR TITLE
Automatically generate the grammar overview

### PR DESCRIPTION
**Pull request by wilzbach**
I guess that tells all?

>  61 additions and 1,851 deletions.

This isn't perfect yet, but already pretty cool.

TODO:
- [ ] https://github.com/dlang/dlang.org/pull/2137

There are 8 new modules which contained D grammar which have been forgotten.

### CODE REVIEW

File: ddoc_preprocessor.d [hunk 1]
```diff
@@ -194,3 +195,61 @@ auto genHeader(string fileText)
...
+        {
+            auto ps = line.splitter(ddocEntryStart).dropOne.front.splitter(",");
+            auto name = ps.front.stripExtension.withExtension(".dd").to!string;
+            auto fileName = specDir.buildPath(name);
+            auto title = ps.dropOne.front.idup.strip;
+            entries ~= Entry(name, title, fileName);
+        }
+    }
+    return entries;
+}
```
Comment (wilzbach) [inline ddoc_preprocessor.d]: This function is copied from the footer generation. I plan to move the footer generation to this script too, see e.g. https://github.com/dlang/dlang.org/pull/2061 (but that isn't a top priority as the footer already exists and we don't add new pages frequently).

File: ddoc_preprocessor.d [hunk 2]
```diff
@@ -194,3 +195,61 @@ auto genHeader(string fileText)
...
+        writeln("Menu file has an invalid format.");
+    foreach (line; specText.splitter("\n"))
+    {
+        enum ddocEntryStart = "$(ROOT_DIR)spec/";
+        if (line.find!(not!isWhite).startsWith(ddocEntryStart))
+        {
+            auto ps = line.splitter(ddocEntryStart).dropOne.front.splitter(",");
+            auto name = ps.front.stripExtension.withExtension(".dd").to!string;
+            auto fileName = specDir.buildPath(name);
+            auto title = ps.dropOne.front.idup.strip;
```
Comment (wilzbach) [inline ddoc_preprocessor.d]: I would love to have [`shift`](https://github.com/dlang/phobos/pull/4010) - doing `front` and `popFront` in one step.

File: ddoc_preprocessor.d [hunk 3]
```diff
@@ -194,3 +195,61 @@ auto genHeader(string fileText)
...
+    import std.uni : toLower;
+
+    enum ddocKey = "$(GRAMMAR_SUMMARY";
+    auto newContent = ddocKey ~ "\n";
+
+    if (fileText.canFind(ddocKey))
+    {
+        foreach (i, entry; specTocEntries)
+        {
+            if (entry.fileName.endsWith("grammar.dd", "lex.dd", "simd.dd"))
```
Comment (wilzbach) [inline ddoc_preprocessor.d]: - `sim.dd` defines this:

```
$(GRAMMAR
$(I typeNN)
)
```

- `lex.dd` is excluded in the current spec

image

**Comment by CyberShadow** [comment]
@wilzbach 
I noticed https://dlang.org/spec/grammar.html has some broken links (e.g. what `__PRETTY_FUNCTION__` links to) and I don't understand why. When I look at the original location for that grammar block, in `lex`, it looks fine (in `lex.verbatim` it shows as `$(LINK2 expression.html#specialkeywords, $(D __PRETTY_FUNCTION__))`). But in `grammar.verbatim` it shows up as `$(D $(RELATIVE_LINK2 specialkeywords, __PRETTY_FUNCTION__))`. I don't know where the `RELATIVE_LINK2` is coming from.

**Pull request merged**